### ISSUE TITLE
Is there an universal way to retrive the value of an atom without using useAtom?

### ISSUE THREAD

**Issue by dzcpy** [pmndrs/jotai#454]
Is there an equivalent way to fetch the atom's value as in `zustand`, using `someStore.getState()`?
I'm currently having some issue to get this work:

```ts
type IdToPathMap = { [key: string]: string }
const idToPathMapAtom = atom<IdToPathMap>({})

type Element = { id: string, payload: any, children: Element[] }
const elementsAtom = atom<Element>([])

export const elementAtomFamily = atomFamily(
  (id) => {
    // How to get the path here?
   const path = idToPathMap[id];
    return focusAtom(elementsAtom, (optic) => optic.path(path))
  },
  isEqual,
);
```

**Comment by dai-shi** [comment]
> Is there an equivalent way to fetch the atom's value as in `zustand`, using `someStore.getState()`?

Short answer is no.

There exists multiple atom values for a single atom config at the same time. So, there's no notion of getting the single atom value just from a config. You have to `useAtom`.
Like `useState`, bridging react state and external variable, you would need `useEffect`.
Or, jotai provides `jotai/zustand` bundle to make the life easier.

> I'm currently having some issue to get this work:
> 
> ```ts
> type IdToPathMap = { [key: string]: string }
> const idToPathMapAtom = atom<IdToPathMap>({})
> 
> type Element = { id: string, payload: any, children: Element[] }
> const elementsAtom = atom<Element>([])
> 
> export const elementAtomFamily = atomFamily(
>   (id) => {
>     // How to get the path here?
>    const path = idToPathMap[id];
>     return focusAtom(elementsAtom, (optic) => optic.path(path))
>   },
>   isEqual,
> );
> ```

You might need to tweak your head to better grasp the abstraction.
Let's simplify assuming you only need a readonly atom family.

```js
export const elementAtomFamily = atomFamily(
  (id) => atom((get) => {
    const idToPathMap = get(idToPathMap)
    const path = idToPathMap[id]
    return lodash.get(get(elementsAtom), path)
  })
);
```

> Another use case is that sometimes I need to retrive the data of an atom in an event handler, without causing rerender when the data changes.

That's when writable/write-only atoms come to play.

Basically, what you are trying to do could be done only with `atom` and `useAtom`.
Even `atomFamily` is unnecessary complication in your use case.

```ts
type Element = { id: string, payload: any, childIds: string[] }
type ElementAtom = PrimitiveAtom<Element>
type IdToElementAtomMap = { [key: string]: ElementAtom }
const idToElementAtomMapAtom = atom<IdToElementAtomMap>({})
```

Something like this would work. But as we discuss in the other issue, you want to serialize it, so big atom pattern might fit better than this.

---

I'm happy to give you more examples to help understand the abstraction. Please give me a small problem for each.

**Comment by dzcpy** [comment]
@dai-shi 
Thanks very much for your reply. I think now I can resolve all of my issues .
At the same time I'm kind of thinking if I could hack it a little bit to save the getter function actually in an atom

Like:

```ts
const intAtom = atom(1)
const getAtom = atom(get => get)
```

```ts
// in React component
const get = useAtomValue(getAtom)
const int = get(intAtom)
```

This seems a bit wierd but I'll see if it works 🙄

**Comment by dai-shi** [comment]
Good try. What you would want for the "write" part is this:

```js
const intAtom = atom(1)
const setAnyAtom = atom(
  null,
  (get, set, [inputAtom, value]) => {
    set(inputAtom, value))
  }
)

  // in React component
  const [, set] = useAtom(setAnyAtom)
  set([intAtom, 2])
```

For (non-tracking) "read", it's tricky, because it's async.
Take a look at `useAsyncCallback`. That's something you want.

But using these two techniques is not "jotai" way. I feel like there's a better pattern for your use case.

**Comment by dai-shi** [comment]
```js
const getAtom = atom(get => get)
```

haha, this is hacky. hm, I used to check this, but not any more. note that even if it works now, it may not in the future. `read` is expected to be a pure function.

**Comment by dzcpy** [comment]
I've tested this hacky workaround and it works alright. So instead of using `useMemo` or `focusAtom` which could use extra memory for caching results, now I can use getter and setter plus custom hooks to do the same thing. And it's much more flexible.

For example:
> ```ts
> type IdToPathMap = { [key: string]: string }
> const idToPathMapAtom = atom<IdToPathMap>({})
> 
> type Element = { id: string, payload: any, children: Element[] }
> const elementsAtom = atom<Element>([])
> 
> export const elementAtomFamily = atomFamily(
>   (id) => {
>     // How to get the path here?
>    const path = idToPathMap[id];
>     return focusAtom(elementsAtom, (optic) => optic.path(path))
>   },
>   isEqual,
> );
> ```
The previous code snippet can be simplified like this:

```ts
// Atoms
type IdToPathMap = { [key: string]: string }
const idToPathMapAtom = atom<IdToPathMap>({})

type Element = { id: string, payload: any, children: Element[] }
const elementsAtom = atom<Element>([])

const hackyAtom = atom(
  get => get,
  (get, set, [inputAtom, value]) => {
    set(inputAtom, value))
  }
)
```

```ts
// Hook
const useElementData = (id) => {
  const [getAtom, setAtom] = useAtom(hackyAtom)
  const elements = useAtomValue(elementsAtom)
  const idToPathMap = getAtom(idToPathMapAtom)
  const path = idToPathMap[id];
  return [
    _.get(elements, path),
    (value) => setAtom([
      elementsAtom,
      produce(elements, draft => {
        _.set(draft, path, value)
      }
    )]
  )]
}
```

```ts
// React component
const [elementData, setElementData] = useElementData('id-1');
```

**Comment by dai-shi** [comment]
Nice try. As I said, `getAtom` is not recommended which may not work in the future.
The recommended way is:
```js
  const path = useAtom(useMemo(() => atom((get) => get(idToPathMapAtom)[id])), [id]))
```

**Comment by dzcpy** [comment]
> ```js
>   const path = useAtom(useMemo(() => atom((get) => get(idToPathMapAtom)[id])), [id]))
> ```

In this case let's say `idToPathMapAtom` got updated, since it's cached based on `id`, will it trigger update? If not `path` would be stale

**Comment by dai-shi** [comment]
> Thanks for your reply.
> 
> > ```js
> >   const path = useAtom(useMemo(() => atom((get) => get(idToPathMapAtom)[id])), [id]))
> > ```
> 
> In this case let's say `idToPathMapAtom` got updated, since it's cached based on `id`, will it trigger an update? If not `path` would be stale then.

With this, it will re-evaluate when `idPathMapAtom` changes and only trigger re-render if `path` changes.

With `getAtom`, it may not trigger re-render when `idPathMapAtom` changes. At least not guaranteed.

> Or should it be more like this?
> 
> ```js
>   const idToPathMap = useAtomValue(idToPathMapAtom)
>   const path = useAtom(useMemo(() => atom((get) => idToPathMapAtom[id])), [id, idToPathMap]))
> ```

With this, it will re-render every time `idToPathMap` is changed, regardless of the case if `path` isn't changed. So, it's not something you want.

> But with this approach, when `idToPathMap` changes very fraquently, the cache might keep growing, it won't cause a huge problem I guess, but it could cause small memory leaks. I'd still prefer there is a better approach, if there is any

Oh, by the way, we take the memory leak issue seriously. We want to fully support these use cases. Implementation-wise, atom values are stored in WeakMaps, so it should be garbage collected. garbage collection is not free though. If you want more control in caching, you can try `atomFamily` or your own cache.
